File: mip_reference/ruby-sinatra/lib/mip/models/node_identity.rb
# frozen_string_literal: true

module MIP
  module Models
    # Represents this node's identity in the MIP network
    class NodeIdentity
      attr_reader :mip_identifier, :private_key, :public_key, :organization_name,
                  :contact_person, :contact_phone, :mip_url, :share_my_organization,
                  :trust_threshold, :port

      def initialize(attrs = {})
        @mip_identifier = attrs[:mip_identifier]
        @private_key = attrs[:private_key]
        @public_key = attrs[:public_key]
        @organization_name = attrs[:organization_name]
        @contact_person = attrs[:contact_person]
        @contact_phone = attrs[:contact_phone]
        @mip_url = attrs[:mip_url]
        @share_my_organization = attrs.fetch(:share_my_organization, true)
        @trust_threshold = attrs.fetch(:trust_threshold, 1)
        @port = attrs[:port]
      end

      def public_key_fingerprint
        MIP::Crypto.fingerprint(@public_key)
      end

      def to_node_profile
        {
          mip_identifier: @mip_identifier,
          mip_url: @mip_url,
          organization_legal_name: @organization_name,
          contact_person: @contact_person,
          contact_phone: @contact_phone,
          public_key: @public_key,
          share_my_organization: @share_my_organization
        }
      end

      # Generate a new node identity from config
      def self.from_config(config, port)
        keys = MIP::Crypto.generate_key_pair
        mip_id = MIP::Identifier.generate(config['organization_name'])

        new(
          mip_identifier: mip_id,
          private_key: keys[:private_key],
          public_key: keys[:public_key],
          organization_name: config['organization_name'],
          contact_person: config['contact_person'],
          contact_phone: config['contact_phone'],
          mip_url: "http://localhost:#{port}/mip/node/#{mip_id}",
          share_my_organization: config.fetch('share_my_organization', true),
          trust_threshold: config.fetch('trust_threshold', 1),
          port: port
        )
      end
    end
  end
end

File: mip_reference/dotnet-aspnet/Mip/Models/NodeIdentity.cs
namespace Mip.Models;

/// <summary>
/// Represents this node's identity in the MIP network.
/// </summary>
public class NodeIdentity
{
    public string MipIdentifier { get; set; } = "";
    public string PrivateKey { get; set; } = "";
    public string PublicKey { get; set; } = "";
    public string OrganizationName { get; set; } = "";
    public string? ContactPerson { get; set; }
    public string? ContactPhone { get; set; }
    public string MipUrl { get; set; } = "";
    public bool ShareMyOrganization { get; set; } = true;
    public int TrustThreshold { get; set; } = 1;
    public int Port { get; set; }

    public string PublicKeyFingerprint => Crypto.Fingerprint(PublicKey);

    public Dictionary<string, object?> ToNodeProfile()
    {
        return new Dictionary<string, object?>
        {
            ["mip_identifier"] = MipIdentifier,
            ["mip_url"] = MipUrl,
            ["organization_legal_name"] = OrganizationName,
            ["contact_person"] = ContactPerson,
            ["contact_phone"] = ContactPhone,
            ["public_key"] = PublicKey,
            ["share_my_organization"] = ShareMyOrganization
        };
    }

    /// <summary>
    /// Generate a new node identity from configuration.
    /// </summary>
    public static NodeIdentity FromConfig(NodeConfig config, int port)
    {
        var keys = Crypto.GenerateKeyPair();
        var mipId = Identifier.Generate(config.OrganizationName);

        return new NodeIdentity
        {
            MipIdentifier = mipId,
            PrivateKey = keys.PrivateKey,
            PublicKey = keys.PublicKey,
            OrganizationName = config.OrganizationName,
            ContactPerson = config.ContactPerson,
            ContactPhone = config.ContactPhone,
            MipUrl = $"http://localhost:{port}/mip/node/{mipId}",
            ShareMyOrganization = config.ShareMyOrganization,
            TrustThreshold = config.TrustThreshold,
            Port = port
        };
    }
}

/// <summary>
/// Configuration loaded from YAML file.
/// </summary>
public class NodeConfig
{
    public int Port { get; set; }
    public string OrganizationName { get; set; } = "";
    public string? ContactPerson { get; set; }
    public string? ContactPhone { get; set; }
    public bool ShareMyOrganization { get; set; } = true;
    public int TrustThreshold { get; set; } = 1;
    public List<MemberConfig> Members { get; set; } = new();
}

public class MemberConfig
{
    public string MemberNumber { get; set; } = "";
    public string? Prefix { get; set; }
    public string FirstName { get; set; } = "";
    public string? MiddleName { get; set; }
    public string LastName { get; set; } = "";
    public string? Suffix { get; set; }
    public string? Honorific { get; set; }
    public string? Rank { get; set; }
    public string? Birthdate { get; set; }
    public int? YearsInGoodStanding { get; set; }
    public string Status { get; set; } = "Active";
    public bool IsActive { get; set; } = true;
    public bool GoodStanding { get; set; } = true;
    public string? Email { get; set; }
    public string? Phone { get; set; }
    public string? Cell { get; set; }
    public AddressConfig? Address { get; set; }
    public List<AffiliationConfig> Affiliations { get; set; } = new();
    public List<LifeCycleEventConfig> LifeCycleEvents { get; set; } = new();
}

public class AddressConfig
{
    public string? Address1 { get; set; }
    public string? Address2 { get; set; }
    public string? City { get; set; }
    public string? State { get; set; }
    public string? PostalCode { get; set; }
    public string? Country { get; set; }
}

public class AffiliationConfig
{
    public string? LocalName { get; set; }
    public string? LocalNumber { get; set; }
    public string? LocalLocation { get; set; }
    public string? MemberType { get; set; }
    public string? LocalStatus { get; set; }
    public bool IsActive { get; set; } = true;
    public string? Level { get; set; }
    public string? Date { get; set; }
}

public class LifeCycleEventConfig
{
    public string? EventName { get; set; }
    public string? Date { get; set; }
    public string? LocalName { get; set; }
}
